Found a data race in controller/metrics

```

==================
WARNING: DATA RACE
Write at 0x00c00048efc0 by goroutine 89:
  runtime.mapassign()
      /usr/local/go/src/runtime/map.go:571 +0x0
  reflect.mapassign()
      /usr/local/go/src/runtime/map.go:1319 +0x3e
  github.com/json-iterator/go.(*structFieldDecoder).Decode()
      /home/hongchao/go/pkg/mod/github.com/json-iterator/go@v1.1.10/reflect_struct_decoder.go:1054 +0xc8
  github.com/json-iterator/go.(*generalStructDecoder).decodeOneField()
      /home/hongchao/go/pkg/mod/github.com/json-iterator/go@v1.1.10/reflect_struct_decoder.go:552 +0x1e0
  github.com/json-iterator/go.(*generalStructDecoder).Decode()
      /home/hongchao/go/pkg/mod/github.com/json-iterator/go@v1.1.10/reflect_struct_decoder.go:508 +0x10a
  github.com/json-iterator/go.(*structFieldDecoder).Decode()
      /home/hongchao/go/pkg/mod/github.com/json-iterator/go@v1.1.10/reflect_struct_decoder.go:1054 +0xc8
  github.com/json-iterator/go.(*fiveFieldsStructDecoder).Decode()
      /home/hongchao/go/pkg/mod/github.com/json-iterator/go@v1.1.10/reflect_struct_decoder.go:735 +0x166
  github.com/json-iterator/go.(*Iterator).ReadVal()
      /home/hongchao/go/pkg/mod/github.com/json-iterator/go@v1.1.10/reflect.go:79 +0x1ea
  github.com/json-iterator/go.(*frozenConfig).Unmarshal()
      /home/hongchao/go/pkg/mod/github.com/json-iterator/go@v1.1.10/config.go:348 +0xf9
  k8s.io/apimachinery/pkg/runtime/serializer/json.(*Serializer).Decode()
      /home/hongchao/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/runtime/serializer/json/json.go:263 +0x549
  k8s.io/apimachinery/pkg/runtime.WithoutVersionDecoder.Decode()
      /home/hongchao/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/runtime/helper.go:252 +0xa7
  k8s.io/apimachinery/pkg/runtime.(*WithoutVersionDecoder).Decode()
      <autogenerated>:1 +0xd9
  k8s.io/client-go/rest.Result.Into()
      /home/hongchao/go/pkg/mod/k8s.io/client-go@v0.18.8/rest/request.go:1233 +0xc4
  sigs.k8s.io/controller-runtime/pkg/client.(*typedClient).Patch()
      /home/hongchao/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/client/typed_client.go:127 +0x802
  sigs.k8s.io/controller-runtime/pkg/client.(*client).Patch()
      /home/hongchao/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/client/client.go:156 +0x27d
  sigs.k8s.io/controller-runtime/pkg/client.(*DelegatingClient).Patch()
      <autogenerated>:1 +0xe6
  github.com/oam-dev/kubevela/pkg/controller/v1alpha1/metrics.(*Reconciler).createService()
      /home/hongchao/go/src/github.com/oam-dev/kubevela/pkg/controller/v1alpha1/metrics/metricstrait_controller.go:221 +0x8e6
  github.com/oam-dev/kubevela/pkg/controller/v1alpha1/metrics.(*Reconciler).Reconcile()
      /home/hongchao/go/src/github.com/oam-dev/kubevela/pkg/controller/v1alpha1/metrics/metricstrait_controller.go:133 +0x240a
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler()
      /home/hongchao/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:233 +0x1cf
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem()
      /home/hongchao/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:209 +0x12f
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker()
      /home/hongchao/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:188 +0x38
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker-fm()
      /home/hongchao/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:187 +0x41
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /home/hongchao/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:155 +0x75
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /home/hongchao/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:156 +0xb3
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /home/hongchao/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:133 +0x10d
  k8s.io/apimachinery/pkg/util/wait.Until()
      /home/hongchao/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:90 +0x5a

Previous read at 0x00c00048efc0 by goroutine 52:
  reflect.maplen()
      /usr/local/go/src/runtime/map.go:1351 +0x0
  reflect.Value.Len()
      /usr/local/go/src/reflect/value.go:1139 +0x2f1
  reflect.deepValueEqual()
      /usr/local/go/src/reflect/deepequal.go:124 +0x1484
  reflect.DeepEqual()
      /usr/local/go/src/reflect/deepequal.go:210 +0x339
  github.com/onsi/gomega/matchers.(*EqualMatcher).Match()
      /home/hongchao/go/pkg/mod/github.com/onsi/gomega@v1.10.1/matchers/equal_matcher.go:27 +0xec
  github.com/onsi/gomega/internal/assertion.(*Assertion).match()
      /home/hongchao/go/pkg/mod/github.com/onsi/gomega@v1.10.1/internal/assertion/assertion.go:64 +0x8f
  github.com/onsi/gomega/internal/assertion.(*Assertion).Should()
      /home/hongchao/go/pkg/mod/github.com/onsi/gomega@v1.10.1/internal/assertion/assertion.go:28 +0x11b
  github.com/oam-dev/kubevela/pkg/controller/v1alpha1/metrics.glob..func1.3()
      /home/hongchao/go/src/github.com/oam-dev/kubevela/pkg/controller/v1alpha1/metrics/metricstrait_controller_integration_test.go:144 +0xf6a
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
      /home/hongchao/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/leafnodes/runner.go:113 +0xf3
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
      /home/hongchao/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/leafnodes/runner.go:64 +0x107
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run()
      /home/hongchao/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
      /home/hongchao/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/spec/spec.go:215 +0x6e1
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
      /home/hongchao/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/spec/spec.go:138 +0x17e
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
      /home/hongchao/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/specrunner/spec_runner.go:200 +0x172
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
      /home/hongchao/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/specrunner/spec_runner.go:170 +0x216
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
      /home/hongchao/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/specrunner/spec_runner.go:66 +0x140
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
      /home/hongchao/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/suite/suite.go:62 +0x69f
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
      /home/hongchao/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/ginkgo_dsl.go:219 +0x338
  github.com/onsi/ginkgo.RunSpecsWithDefaultAndCustomReporters()
      /home/hongchao/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/ginkgo_dsl.go:207 +0x11f
  github.com/oam-dev/kubevela/pkg/controller/v1alpha1/metrics.TestAPIs()
      /home/hongchao/go/src/github.com/oam-dev/kubevela/pkg/controller/v1alpha1/metrics/suite_test.go:57 +0x138
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1039 +0x1eb

Goroutine 89 (running) created at:
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
      /home/hongchao/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:170 +0x5fc
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start()
      /home/hongchao/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:175 +0x154
  sigs.k8s.io/controller-runtime/pkg/manager.(*controllerManager).startLeaderElectionRunnables.func1()
      /home/hongchao/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/manager/internal.go:514 +0x83

Goroutine 52 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1090 +0x700
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1334 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1039 +0x1eb
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1332 +0x527
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1249 +0x43f
  main.main()
      _testmain.go:92 +0x337
==================
••

Ran 2 of 2 Specs in 31.475 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 0 Skipped
--- FAIL: TestAPIs (31.48s)
    testing.go:954: race detected during execution of test
FAIL
coverage: 57.0% of statements
FAIL	github.com/oam-dev/kubevela/pkg/controller/v1alpha1/metrics	31.571s
```

--- a/pkg/controller/v1alpha1/metrics/metricstrait_controller.go
+++ b/pkg/controller/v1alpha1/metrics/metricstrait_controller.go
@@ -55,16 +55,19 @@ var (
 )
 
 var (
-	// oamServiceLabel is the pre-defined labels for any serviceMonitor
-	// created by the MetricsTrait,  prometheus operator listens on this
-	oamServiceLabel = map[string]string{
-		"k8s-app":    "oam",
-		"controller": "metricsTrait",
-	}
 	// ServiceMonitorNSName is the name of the namespace in which the serviceMonitor resides
 	// it must be the same that the prometheus operator is listening to
 	ServiceMonitorNSName = "monitoring"
 )
+
+// GetOAMServiceLabel will return oamServiceLabel as the pre-defined labels for any serviceMonitor
+// created by the MetricsTrait,  prometheus operator listens on this
+func GetOAMServiceLabel() map[string]string {
+	return map[string]string{
+		"k8s-app":    "oam",
+		"controller": "metricsTrait",
+	}
+}
 
 // Reconciler reconciles a MetricsTrait object
 type Reconciler struct {
@@ -200,7 +203,7 @@ func (r *Reconciler) createService(ctx context.Context, mLog logr.Logger, worklo
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      "oam-" + workload.GetName(),
 			Namespace: workload.GetNamespace(),
-			Labels:    oamServiceLabel,
+			Labels:    GetOAMServiceLabel(),
 			OwnerReferences: []metav1.OwnerReference{
 				{
 					APIVersion:         metricsTrait.GetObjectKind().GroupVersionKind().GroupVersion().String(),
@@ -291,7 +294,7 @@ func constructServiceMonitor(metricsTrait *v1alpha1.MetricsTrait, targetPort int
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      metricsTrait.Name,
 			Namespace: ServiceMonitorNSName,
-			Labels:    oamServiceLabel,
+			Labels:    GetOAMServiceLabel(),
 			OwnerReferences: []metav1.OwnerReference{
 				{
 					APIVersion:         metricsTrait.GetObjectKind().GroupVersionKind().GroupVersion().String(),
@@ -305,7 +308,7 @@ func constructServiceMonitor(metricsTrait *v1alpha1.MetricsTrait, targetPort int
 		},
 		Spec: monitoring.ServiceMonitorSpec{
 			Selector: metav1.LabelSelector{
-				MatchLabels: oamServiceLabel,
+				MatchLabels: GetOAMServiceLabel(),
 			},
 			// we assumed that the service is in the same namespace as the trait
 			NamespaceSelector: monitoring.NamespaceSelector{

--- a/pkg/controller/v1alpha1/metrics/metricstrait_controller_integration_test.go
+++ b/pkg/controller/v1alpha1/metrics/metricstrait_controller_integration_test.go
@@ -141,7 +141,7 @@ var _ = Describe("Metrics Trait Integration Test", func() {
 			time.Second*10, time.Millisecond*500).Should(BeNil())
 		logf.Log.Info("[TEST] Get the created service", "service ports", createdService.Spec.Ports)
 		Expect(createdService.GetNamespace()).Should(Equal(namespaceName))
-		Expect(createdService.Labels).Should(Equal(oamServiceLabel))
+		Expect(createdService.Labels).Should(Equal(GetOAMServiceLabel()))
 		Expect(len(createdService.Spec.Ports)).Should(Equal(1))
 		Expect(createdService.Spec.Ports[0].Port).Should(BeEquivalentTo(servicePort))
 		Expect(createdService.Spec.Selector).Should(Equal(deployLabel))
@@ -156,7 +156,7 @@ var _ = Describe("Metrics Trait Integration Test", func() {
 			time.Second*5, time.Millisecond*50).Should(BeNil())
 		logf.Log.Info("[TEST] Get the created serviceMonitor", "service end ports", serviceMonitor.Spec.Endpoints)
 		Expect(serviceMonitor.GetNamespace()).Should(Equal(ServiceMonitorNSName))
-		Expect(serviceMonitor.Spec.Selector.MatchLabels).Should(Equal(oamServiceLabel))
+		Expect(serviceMonitor.Spec.Selector.MatchLabels).Should(Equal(GetOAMServiceLabel()))
 		Expect(serviceMonitor.Spec.Selector.MatchExpressions).Should(BeNil())
 		Expect(serviceMonitor.Spec.NamespaceSelector.MatchNames).Should(Equal([]string{metricsTrait.Namespace}))
 		Expect(serviceMonitor.Spec.NamespaceSelector.Any).Should(BeFalse())
@@ -192,7 +192,7 @@ var _ = Describe("Metrics Trait Integration Test", func() {
 			},
 			time.Second*10, time.Millisecond*500).Should(BeNil())
 		logf.Log.Info("[TEST] Get the created service", "service ports", createdService.Spec.Ports)
-		Expect(createdService.Labels).Should(Equal(oamServiceLabel))
+		Expect(createdService.Labels).Should(Equal(GetOAMServiceLabel()))
 		Expect(createdService.Spec.Selector).Should(Equal(deployLabel))
 		By("Check that we have created the serviceMonitor in the pre-defined namespaceName")
 		var serviceMonitor monitoringv1.ServiceMonitor
@@ -204,7 +204,7 @@ var _ = Describe("Metrics Trait Integration Test", func() {
 			},
 			time.Second*5, time.Millisecond*50).Should(BeNil())
 		logf.Log.Info("[TEST] Get the created serviceMonitor", "service end ports", serviceMonitor.Spec.Endpoints)
-		Expect(serviceMonitor.Spec.Selector.MatchLabels).Should(Equal(oamServiceLabel))
+		Expect(serviceMonitor.Spec.Selector.MatchLabels).Should(Equal(GetOAMServiceLabel()))
 		Expect(serviceMonitor.Spec.Selector.MatchExpressions).Should(BeNil())
 		Expect(serviceMonitor.Spec.NamespaceSelector.MatchNames).Should(Equal([]string{metricsTrait.Namespace}))
 		Expect(serviceMonitor.Spec.NamespaceSelector.Any).Should(BeFalse())

